### ISSUE TITLE
Unexpected non-scalars in outputs of `asm!` cause rustc to dump core

### ISSUE THREAD

**Issue by cbiffle** [rust-lang/rust#37435]
Unexpected non-scalars in outputs of `asm!` cause rustc to dump core.  (Doing this inputs causes a different problem; see #37433).

To reproduce:

``` rust
#![feature(asm)]

fn main() {
    unsafe {
        asm!("" : "=r"(""));
    }
}
```

I get:

```
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/lib/IR/InlineAsm.cpp:46: llvm::InlineAsm::InlineAsm(llvm::FunctionType*, const string&, const string&, bool, bool, llvm::InlineAsm::AsmDialect): Assertion `Verify(getFunctionType(), constraints) && "Function type not legal for constraints!"' failed.
Aborted (core dumped)
```
## Meta

`rustc --version --verbose`:
rustc 1.14.0-nightly (c59cb71d9 2016-10-26)
binary: rustc
commit-hash: c59cb71d976ceabf00c7da0224a795fab530601e
commit-date: 2016-10-26
host: x86_64-unknown-linux-gnu
release: 1.14.0-nightly
LLVM version: 3.9

No backtrace was produced (core dumped).

**Comment by arielb1** [comment]
This is an LLVM assertion:

```
rustc: /home/ariel/Rust/rust-master/src/llvm/lib/IR/InlineAsm.cpp:46: llvm::InlineAsm::InlineAsm(llvm::FunctionType*, const string&, const string&, bool, bool, llvm::InlineAsm::AsmDialect): Assertion `Verify(getFunctionType(), constraints) && "Function type not legal for constraints!"' failed.
Aborted
```

**Comment by levex** [comment]
No longer reproduces on `nightly`: https://play.rust-lang.org/?gist=ff40c68f35055c97aaa471149cc23e37&version=nightly&mode=debug&edition=2015

(Fixed by PR #54568)

**Comment by Amanieu** [comment]
This issue does not apply to the new `asm!` ([RFC 2850](https://github.com/rust-lang/rfcs/pull/2850)) which has stricter checks on the types that can be used as asm operands.

The legacy `llvm_asm!` is deprecated and is no longer maintained.
